### ISSUE TITLE
Backspacing a new line on the web causes NVDA to read the character prior to the new line

### ISSUE THREAD

**Issue by Qchristensen** [nvaccess/nvda#12375]
### Steps to reproduce:
1. Open a website with a multi-line edit.  (I can recreate in Gmail, Google Docs, Twitter and Facebook, in Chrome and Firefox, although interestingly, not GitHub).
2. Type at least one character and a newline (enter)
3. Backspace to remove the newline
4. Backspace to remove a character

### Actual behavior:

At step 3, NVDA removes the newline, moves to after the last character on the line above but reads that last character on the line above.

At step 4, NVDA reads the same character, as that is the character it just deleted.

### Expected behavior:

NVDA should read the character deleted in each instance (it should read the newline or linefeed character at step 3, not the character to the left of the caret now).

#### NVDA version:

NVDA 2020.4

#### Windows version:

Windows 10 (64-bit) Version: 2009, Build: 19042 

#### Name and version of other software in use when reproducing the issue:

Firefox version: 88.0 
Chrome Version: 90.0.4430.93 

#### Does the issue still occur after restarting your computer?

Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

No

#### If add-ons are disabled, is your problem still occurring?

Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

**Comment by JulienCochuyt** [comment]
I do reproduce on GMail in Firefox, but not in Chrome 90.0.4430.93 under Windows 10 20H2 19042.964 with a French locale.
Here, at step 3:
 - In Notepad++, NVDA says "blank": KO
 - In Notepad, NVDA says "new line": OK
 - In Firefox / GMail, NVDA says the previous character: KO
 - In Chrome / GMail, NVDA says "new line": OK

**Comment by JulienCochuyt** [comment]
The issue seems to come from the fact NVDA cannot fetch the actual caret position while in focus mode in Chrome nor Firefox.
Hence, the "announce deleted character" code is bypassed and the announce is solely the result of the next events received after letting the `backspace` key through.
The "working" cases seem to be the result of different events being sent.
Try by example to select a couple character in the multi-line text entry on GitHub and hit `backspace`.
Now do the same in the Windows Notepad.
The resulting announce is different.

As for the issue I noted in Notepad++, this one comes from the fact Scintilla TextInfo carries "`\r\n`" instead of "`\n`" and NVDA treats it as the deletion of two characters, thus fails to announce it properly.
I'm preparing a PR to fix this one.

There is also another issue in the handling of `control+backspace`:
 - When there is no selection, the whole word is announce instead of the deleted chunk (noticeable if you hit `control+backspace` while the caret is in the middle of the word). I'll try to address this one too.
 - When there is a selection, however, different editors behave very differently. I guess the only truly working solution would be to always first let the key pass through, and inspect the actual result instead of trying to be clever and and determine it in advance.

**Comment by ainekc** [comment]
A related issue is that when moving by paragraph past a blank line, NVDA announces the paragraph above. I imagine the source of the problem could be the same? Although this one *does* also occur in Github.

**Comment by Adriani90** [comment]
@Qchristensen can you still reproduce this issue now after #12379 has been merged?

**Comment by Qchristensen** [comment]
I just tried here in the comment field on GitHub in Chrome Version: 112.0.5615.138  with NVDA 2023.1.0.27913 on Windows 11 (64-bit) Version: 22H2, Build: 22621.1555 

and when you backspace the line feed, NVDA correctly reports it as such:

Input: kb(desktop):backspace
DEBUG - editableText.EditableText._hasCaretMoved (11:26:31.027) - MainThread (12848):
Caret move detected using bookmarks. Elapsed: 0 ms
IO - speech.speech.speak (11:26:31.027) - MainThread (12848):
Speaking [LangChangeCommand ('en_GB'), 'line feed', EndUtteranceCommand()]

But when I tried on Gmail, I got the same behaviour as previously (Reading the incorrect character):

Input: kb(desktop):backspace
DEBUG - editableText.EditableText._hasCaretMoved (11:26:49.135) - MainThread (12848):
Caret move detected using bookmarks. Elapsed: 10 ms
IO - speech.speech.speak (11:26:49.135) - MainThread (12848):
Speaking [CharacterModeCommand(True), LangChangeCommand ('en_GB'), 'd', EndUtteranceCommand()]

I don't know if the 10ms elapsed time has anything to do with it trying several things - or just that it takes around 5ms and 10 is rounded up and 0 is rounded down?

**Comment by Adriani90** [comment]
Do you get the same behavior with NVDA last alpha version?Von meinem iPhone gesendetAm 26.04.2023 um 03:32 schrieb Quentin Christensen ***@***.***>:﻿
I just tried here in the comment field on GitHub in Chrome Version: 112.0.5615.138  with NVDA 2023.1.0.27913 on Windows 11 (64-bit) Version: 22H2, Build: 22621.1555
and when you backspace the line feed, NVDA correctly reports it as such:
Input: kb(desktop):backspace
DEBUG - editableText.EditableText._hasCaretMoved (11:26:31.027) - MainThread (12848):
Caret move detected using bookmarks. Elapsed: 0 ms
IO - speech.speech.speak (11:26:31.027) - MainThread (12848):
Speaking [LangChangeCommand ('en_GB'), 'line feed', EndUtteranceCommand()]
But when I tried on Gmail, I got the same behaviour as previously (Reading the incorrect character):
Input: kb(desktop):backspace
DEBUG - editableText.EditableText._hasCaretMoved (11:26:49.135) - MainThread (12848):
Caret move detected using bookmarks. Elapsed: 10 ms
IO - speech.speech.speak (11:26:49.135) - MainThread (12848):
Speaking [CharacterModeCommand(True), LangChangeCommand ('en_GB'), 'd', EndUtteranceCommand()]
I don't know if the 10ms elapsed time has anything to do with it trying several things - or just that it takes around 5ms and 10 is rounded up and 0 is rounded down?

**Comment by Qchristensen** [comment]
Interesting, I just tried with Version: alpha-28130,a0f16e0b (2023.2.0.28130)
and it works fine here, but in GMail, it would backspace as many letters as you like on the current line correctly, but when I backspaced a line feed it read the last character on the previous line, and then wouldn't read any other characters I backspaced on that line.

Removing the linefeed:

IO - inputCore.InputManager.executeGesture (14:54:14.267) - winInputHook (11028):
Input: kb(desktop):backspace
DEBUG - editableText.EditableText._hasCaretMoved (14:54:14.326) - MainThread (24308):
Caret move detected using bookmarks. Elapsed 0.00210309 sec, retries 4
IO - speech.speech.speak (14:54:14.327) - MainThread (24308):
Speaking [CharacterModeCommand(True), LangChangeCommand ('en_GB'), 'k', EndUtteranceCommand()]

Backspace again:

IO - inputCore.InputManager.executeGesture (14:54:14.898) - winInputHook (11028):
Input: kb(desktop):backspace
DEBUG - editableText.EditableText._hasCaretMoved (14:54:15.019) - MainThread (24308):
Caret didn't move before timeout. Elapsed: 0.108882 sec

Visually everything moved as normal.

**Comment by Adriani90** [comment]
Cc: @jcstehVon meinem iPhone gesendetAm 26.04.2023 um 06:59 schrieb Quentin Christensen ***@***.***>:﻿
Interesting, I just tried with Version: alpha-28130,a0f16e0b (2023.2.0.28130)
and it works fine here, but in GMail, it would backspace as many letters as you like on the current line correctly, but when I backspaced a line feed it read the last character on the previous line, and then wouldn't read any other characters I backspaced on that line.
Removing the linefeed:
IO - inputCore.InputManager.executeGesture (14:54:14.267) - winInputHook (11028):
Input: kb(desktop):backspace
DEBUG - editableText.EditableText._hasCaretMoved (14:54:14.326) - MainThread (24308):
Caret move detected using bookmarks. Elapsed 0.00210309 sec, retries 4
IO - speech.speech.speak (14:54:14.327) - MainThread (24308):
Speaking [CharacterModeCommand(True), LangChangeCommand ('en_GB'), 'k', EndUtteranceCommand()]
Backspace again:
IO - inputCore.InputManager.executeGesture (14:54:14.898) - winInputHook (11028):
Input: kb(desktop):backspace
DEBUG - editableText.EditableText._hasCaretMoved (14:54:15.019) - MainThread (24308):
Caret didn't move before timeout. Elapsed: 0.108882 sec
Visually everything moved as normal.

**Comment by jcsteh** [comment]
This occurs because there's actually no line feed character in this case. Instead, you're dealing with two paragraphs and paragraphs do not end with a line feed character. For example, here's a text box with two paragraphs:
`data:text/html,<div role="textbox" aria-multiline="true" contenteditable><p>a</p><p>b</p>`
If you go to the start of the second line and press backspace, NVDA incorrectly says "a".
Here's a text box containing two lines with a forced line break:
`data:text/html,<textarea>a%0ab</textarea>`
If you go to the start of the second line and press backspace, NVDA says "line feed".

Firefox and Chrome technically aren't doing anything wrong here. That said, I'm not quite sure how to fix this in NVDA.

What does Microsoft Word say if you type a, enter, then backspace? Does it say "carriage return", "line feed", "blank" or something else? (I don't have Word installed here.)

**Comment by XLTechie** [comment]
@jcsteh Here's an abbreviated log from Word 2019, backspacing through the newline and the A:

```
Input: kb(desktop):backspace
Caret move detected using bookmarks. Elapsed 0 sec, retries 0
Speaking ['carriage return', EndUtteranceCommand()]
Not using UIA in MS Word on pre Windows 11 OS due to missing custom extensions
Input: kb(desktop):backspace
Caret move detected using bookmarks. Elapsed 0 sec, retries 0
Speaking [PitchCommand(offset=30), 'A', PitchCommand(), EndUtteranceCommand()]
```

**Comment by Adriani90** [comment]
It seems only browsers are impacted by this, in all text editors I've tested no matter how I add a new line (ctrl+enter, shift+enter, ctrl+shift+enter or only enter), NVDA reads "linefeed" when pressing backspace.
I tested in Notepad, MS Word, notepad++, and Libreoffice writer.

In LO writer, NVDA gives an error and doesn't read the deleted linefeed if it is added by pressing enter.
```
IO - inputCore.InputManager.executeGesture (21:30:01.204) - winInputHook (20560):
Input: kb(laptop):backspace
DEBUGWARNING - NVDAObjects.IAccessible.IA2TextTextInfo._getCaretOffset (21:30:01.212) - MainThread (6368):
IAccessibleText::caretOffset failed
Traceback (most recent call last):
  File "NVDAObjects\IAccessible\__init__.pyc", line 313, in _getCaretOffset
_ctypes.COMError: (-2147467259, 'Unbekannter Fehler', (None, None, None, 0, None))
ERROR - scriptHandler.executeScript (21:30:01.213) - MainThread (6368):
error executing script: <bound method EditableText.script_caret_backspaceCharacter of <appModules.soffice.SymphonyDocument object at 0x0000022B81946A50>> with gesture 'Rücktaste'
Traceback (most recent call last):
  File "NVDAObjects\IAccessible\__init__.pyc", line 313, in _getCaretOffset
_ctypes.COMError: (-2147467259, 'Unbekannter Fehler', (None, None, None, 0, None))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "scriptHandler.pyc", line 300, in executeScript
  File "editableText.pyc", line 324, in script_caret_backspaceCharacter
  File "appModules\soffice.pyc", line 509, in _backspaceScriptHelper
  File "documentBase.pyc", line 76, in makeTextInfo
  File "compoundDocuments.pyc", line 288, in __init__
  File "documentBase.pyc", line 76, in makeTextInfo
  File "textInfos\offsets.pyc", line 500, in __init__
  File "NVDAObjects\IAccessible\__init__.pyc", line 316, in _getCaretOffset
RuntimeError: Retrieving caret offset failed
```

@michaelweghorn could you have a look at the LO part at least? I am using LO writer Version: 26.2.0.0.alpha0+ (X86_64) / LibreOffice community
Build ID: b1c0c4838d2e006ffa8e72516569ce8d13bdbe01

**Comment by Adriani90** [comment]
Note the LO error occurs only when having two text passages with a linefeed between them, like this:
```
Hello
World
```

Removing the linefeed between hello and world gives the error sound in LO.

**Comment by michaelweghorn** [comment]
> [@michaelweghorn](https://github.com/michaelweghorn) could you have a look at the LO part at least? I am using LO writer Version: 26.2.0.0.alpha0+ (X86_64) / LibreOffice community Build ID: b1c0c4838d2e006ffa8e72516569ce8d13bdbe01

@Adriani90 Can you please create a separate issue to keep track of this? (This will need further analysis, which I can't do right now, but will hopefully get to another time.)
